ci: start setting up doctests

nstarman opened the pull request:
This is only a start. Please feel free to edit this PR.

The complications are:

1. running the tessellation analysis takes a very long time. Too long for sequential testing in CI.
2. the figure generation is not yet testable, so there's no way to know if the example code actually generates the associated figure.
3. only the galpy stuff is currently tested.

All of these are solvable. Here's my thoughts how:

1. The in the CI the doctests can be split into a few parallel runners targeting different subsets of the docs by changing the `pytest docs` command to `pytest docs/<subset>`. This should fix the run time.
2. Even then, not all doctests need to be run for every push of a PR as there's a lot of overlap between examples in the docs. But the full functionality should be tested in CI, which it is not currently. So a few tests can be added to `tests/` and then portions of the doctests can be run on a weekly / monthly cron job. If a test fails in `tests/` then it says that a documentation example probably needs to be updated as well. The cron job will eventually catch if it wasn't.
3. Again, not for every PR, but the figures should be dynamically generated. This can be a weekly / monthly cron job. See `ipykernel` or `jupyter` plugins for mkdocs.
4. Using the [sybil's conditional skips](https://sybil.readthedocs.io/en/latest/markdown.html#skipping-examples) the doc tests can be run for each of the three supported backends, not just galpy, in different parts of a CI matrix.
5. Apply the above to the pre-release tests, so that released versions of this package have working docs.

I know this is a fair bit of effort, but IMO being able to run the examples in the docs is the most important test of a package. Users will copy-paste docs examples and they must work. When drafting this PR I found (and fixed) a few non-functional examples; these tests aren't being run on all files, so IDK if I caught everything.

— https://github.com/openjournals/joss-reviews/issues/7009

nstarman committed:
ci: ensure astropy installed in test env

Signed-off-by: nstarman <nstarman@users.noreply.github.com>

nstarman committed:
ci: run pre-commit

Signed-off-by: nstarman <nstarman@users.noreply.github.com>

nstarman committed:
feat: doctests

Signed-off-by: nstarman <nstarman@users.noreply.github.com>

nstarman commented on conftest.py:
```diff
@@ -0,0 +1,32 @@
+
+
+# Ensure pidgey orbit integration works when parsing angles
+u.add_enabled_equivalencies(u.dimensionless_angles())
```
Might want to move this to the docs. The error is unhelpful and most users will not know they need to enable this unit equivalency.

nstarman committed:
ci: add sybil to test

Signed-off-by: nstarman <nstarman@users.noreply.github.com>

nstarman committed:
ci: try putting sybil in dev.

Signed-off-by: nstarman <nstarman@users.noreply.github.com>

nstarman commented:
Hm. IDK why sybil isn't being installed. `tox` installs the `test` dependency-group, but it's not being picked up. I'm trying `dev`, but that doesn't seem to be its intended group.
It does work locally.

ilikecubesnstuff committed:
:pushpin: Update lock file & remove redundant dependencies.

ilikecubesnstuff commented:
> Hm. IDK why sybil isn't being installed. `tox` installs the `test` dependency-group, but it's not being picked up. I'm trying `dev`, but that doesn't seem to be its intended group. It does work locally.

I suppose it was because the lock file wasn't updated to reflect the changes in `pyproject.toml`—it seems to work fine now!

ilikecubesnstuff commented:
Thank you so much for the detailed help with implementing doctests! I have experimented with `mkdocs-jupyter` locally and given this some thought. My main roadblock to progress is generating all the figures dynamically.

The main showcase in much of the documentation is the interactive window with the side-by-side plot. This includes having to click on the graph area to plot an orbit, and the docs have screenshots with manually drawn lines to highlight certain commensurate tracks—I am not sure how to dynamically generate these for arbitrary plots.

The best solution I have thought of is to track a pre-computed HDF5 file of the expected outputs, which would match those in the docs, in version control. Then when the full doctests are being run, a test can be included to check if the docs-generated output matches the precomputed output. This will not generate any new figures, but it will flag when a certain documentation page is invalid so I can tag/amend it accordingly.

ilikecubesnstuff committed:
:memo: Add invisible code blocks and conditional skips to documentation.
Set up skips to be sensitive to the available backend installed.
Mark portions of documentation that need changing/updating.

ilikecubesnstuff committed:
:arrow_up: Update depedency groups and lock file.
Move astropy into required dependencies.
Remove "doctests" group, combine into tests - sybil is required for tests regardless.

ilikecubesnstuff committed:
:construction_worker: Add cron-based testing workflows.

ilikecubesnstuff committed:
:construction_worker: Update tests workflow.
Use miniconda for setup, to match the cron-based workflows.

ilikecubesnstuff committed:
:pushpin: Fix github dependency group and update lock file.

ilikecubesnstuff committed:
:construction_worker: Trigger doctests for PRs with documentation label.

ilikecubesnstuff committed:
:arrow_down: Downgrade astropy dependency.

ilikecubesnstuff committed:
:construction_worker: Run doctests on Python 3.12.

ilikecubesnstuff commented:
Almost all of the workflow setup for automated doctests is complete—the only thing to be added is checks for the figures in the documentation.

That and the remaining documentation changes will be in a separate pull request.
